java/ql/lib/semmle/code/java/controlflow/Guards.qll

```diff
@@ -175,13 +193,74 @@ class Guard extends ExprParent {
   }
 }
 
+/**
+ * A `Guard` that tests an expression's type -- that is, an `instanceof T` or a
+ * `case T varname` pattern case.
+ */
+class TypeTestGuard extends Guard {
+  Expr testedExpr;
+  Type testedType;
+
+  TypeTestGuard() {
+    exists(InstanceOfExpr ioe | this = ioe |
+      testedExpr = ioe.getExpr() and
+      testedType = ioe.getSyntacticCheckedType()
+    )
+    or
+    exists(PatternCase pc | this = pc |
+      pc.getSelectorExpr() = testedExpr and
+      testedType = pc.getPattern().getType()
+    )
+  }
+
+  /**
+   * Gets the record pattern this type test binds to, if any.
```

Comment:
Not necessarily a record
```suggestion
   * Gets the pattern this type test binds to, if any.
```

Reply:
Or maybe just delete this predicate? I don't see any use of it.

Reply:
In fact, this leaves just the charpred and the `appliesTypeTest`, so maybe get rid of the class and just have an `isTypeTest` predicate directly on `Guard` similar to the `isEquality`?